### PR TITLE
Use c10::variant for enums in C++ API

### PR DESCRIPTION
**Pull request by yf225**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #27933 Use c10::variant-based enums for Nonlinearity and FanMode
* **#26837 Use c10::variant for enums in C++ API**
* #26836 Add std::variant backport as c10::variant

The discussion in https://github.com/pytorch/pytorch/issues/15149 captures it the best, and there are two main benefits of using `c10::variant`:
1. Compared to defining a common enum class for all functions, using `c10::variant` we can constrain the the range of acceptable enums for a particular function.
2. Compared to using `std::string` as enum, using `c10::variant` we can save the string comparison overhead, and also constrain the the range of acceptable enums for a particular function.

Differential Revision: [D17579438](https://our.internmc.facebook.com/intern/diff/D17579438)

### CODE REVIEW

**Comment by yf225** [comment]
@ShahriarSS Would you like to review this PR?

**Comment by ShahriarSS** [comment]
@yf225 I think you forgot about [this one](https://github.com/pytorch/pytorch/blob/b5d15315d88e8e8b2a51b4f53fce9b519ebbb58b/torch/csrc/api/include/torch/nn/options/rnn.h#L37) and [this one](https://github.com/pytorch/pytorch/blob/b5d15315d88e8e8b2a51b4f53fce9b519ebbb58b/torch/csrc/api/include/torch/nn/modules/rnn.h#L37) (Although the second one is in detail namespace so maybe it's not that important).

**Comment by yf225** [comment]
@ShahriarSS Yes those are skipped intentionally because we likely need to rewrite how RNN are implemented in C++ API. I will fix the enum issue when I work on parity for the RNN layers.

**Comment by ShahriarSS** [comment]
@yf225 Then this looks good to me. Just one thing: Is it a bad idea to have a macro that defines the enum and it's extern const (is it even possible)?

**Comment by yf225** [comment]
@ShahriarSS Using macro is a great idea - I will add it in this PR.

**Comment by smessmer** [comment]
Can you give some background on why we want to do this? `enum class` seems like less effort to use.

**Comment by yf225** [comment]
I think the discussion in https://github.com/pytorch/pytorch/issues/15149 captures it the best, and I think there are three main benefits of using `std::variant`:
1. Compared to defining an enum class for each function, using `std::variant` we can declare the range of acceptable enums inline with the function declaration, which makes the code easier to read.
2. Compared to defining a common enum class for all functions, using `std::variant` we can constrain the the range of acceptable enums for a particular function.
3. Compared to using `std::string` as enum, using `std::variant` we can save the string comparison overhead, and also constrain the the range of acceptable enums for a particular function.

I will update the PR description to reflect this rationale.

File: torch/csrc/api/include/torch/enum.h [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+TORCH_ENUM_DECLARE(Tanh)
+TORCH_ENUM_DECLARE(ReLU)
+TORCH_ENUM_DECLARE(LeakyReLU)
+TORCH_ENUM_DECLARE(FanIn)
+TORCH_ENUM_DECLARE(FanOut)
+
+namespace torch {
+
+/// Variable of `Nonlinearity` type can take one of the following values:
+/// - `torch::kLinear`
```
Comment (ShahriarSS) [inline torch/csrc/api/include/torch/enum.h]: @yf225 Aren't these comments redundant? the accepted values are defined in the body of declaration itself.

File: torch/csrc/api/include/torch/enum.h [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+  enumtype::ConvTranspose2D,
+  enumtype::ConvTranspose3D,
+  enumtype::Sigmoid,
+  enumtype::Tanh,
+  enumtype::ReLU,
+  enumtype::LeakyReLU
+>;
+
+/// Variable of `FanMode` type can take one of the following values:
+/// - `torch::kFanIn`
```
Comment (ShahriarSS) [inline torch/csrc/api/include/torch/enum.h]: Same thing here about being redundant.

File: torch/csrc/api/include/torch/enum.h (hunk 1 above)
Comment (yf225) [reply torch/csrc/api/include/torch/enum.h]: I am a bit worried that people won't directly see the mapping from `enumtype::Linear` to `torch::kLinear`, and this comment should help them discover that.

Comment (yf225) [reply torch/csrc/api/include/torch/enum.h]: @ShahriarSS I changed the enum type from `enumtype::Linear` to `enumtype::torch::kLinear`, so that people can more easily figure out that they should use `torch::kLinear`.

File: torch/csrc/api/include/torch/nn/init.h [hunk 3]
```diff
@@ -1,27 +1,32 @@
...
-  ConvTranspose3D,
-  Sigmoid,
-  Tanh,
-  ReLU,
-  LeakyReLU
-};
-
-enum class FanMode { FanIn, FanOut };
+using Nonlinearity = c10::variant<
+  enumtype::torch::kLinear,
```
Comment (ShahriarSS) [inline torch/csrc/api/include/torch/nn/init.h]: @yf225 ```enumtype::torch::``` doesn't seem right. Why are you using torch namespace both outside and inside ```enumtype```? I think that the inner ```torch``` is unnecessary.

Comment (yf225) [reply torch/csrc/api/include/torch/nn/init.h]: Sounds good, I changed it to `enumtype::kLinear`.

File: torch/csrc/api/include/torch/nn/init.h [hunk 4]
```diff
@@ -1,27 +1,32 @@
...
-  ConvTranspose2D,
-  ConvTranspose3D,
-  Sigmoid,
-  Tanh,
-  ReLU,
-  LeakyReLU
-};
-
-enum class FanMode { FanIn, FanOut };
+using Nonlinearity = c10::variant<
```
Comment (ShahriarSS) [inline torch/csrc/api/include/torch/nn/init.h]: @yf225 This PR looks good. Just one comment: Do you think that it's good idea to declare the variants where they are used (and not at the top of the file)?

**Comment by smessmer** [comment]
@yf225 as I stated in the discussion thread you linked, I'm not convinced this is a good idea. This relies on a central list of "keywords" that can be used in enums and seems unnecessarily complicated. Also, say you have a "dtype" enum and a lot of function taking it. Then, you add a new dtype. Now you need to change all function signatures? Seems like a recipe for inconsistency.

> Compared to defining an enum class for each function, using c10::variant we can declare the range of acceptable enums inline with the function declaration, which makes the code easier to read.

Why not define an `enum class` in the line right before the function declaration? That would have the same effect that you can directly see the possible values when looking at the function declaration.

> Compared to defining a common enum class for all functions, using c10::variant we can constrain the the range of acceptable enums for a particular function.

I don't propose a common enum class, I propose a separate enum class for each use case, possibly for each function.

**Comment by yf225** [comment]
@smessmer 

> @yf225 as I stated in the discussion thread you linked, I'm not convinced this is a good idea. This relies on a central list of "keywords" that can be used in enums and seems unnecessarily complicated. Also, say you have a "dtype" enum and a lot of function taking it. Then, you add a new dtype. Now you need to change all function signatures? Seems like a recipe for inconsistency.

If we have a `dtype` enum that a lot of functions take (and they don't deviate on the range of acceptable values), we can declare a common variant type, like `Nonlinearity` and `FanMode` in this PR.

> > Compared to defining an enum class for each function, using c10::variant we can declare the range of acceptable enums inline with the function declaration, which makes the code easier to read.
> 
> Why not define an `enum class` in the line right before the function declaration? That would have the same effect that you can directly see the possible values when looking at the function declaration.

Agreed that both `enum class` and variant type works for the UX point here.

> > Compared to defining a common enum class for all functions, using c10::variant we can constrain the the range of acceptable enums for a particular function.
> 
> I don't propose a common enum class, I propose a separate enum class for each use case, possibly for each function.

I think the main issue with having a separate enum class for each use case is the user has to type many more characters to use the enum. For example:
```cpp
enum class MSELossReduction {None, Mean, Sum};
Tensor MSELoss(MSELossReduction reduction, ...);

enum class CrossEntropyLossReduction {None, Mean, Sum};
Tensor CrossEntropyLoss(CrossEntropyLossReduction reduction, ...);
```
when the user calls MSELoss with reduction `None`, they will need to write 
```
MSELoss(torch::MSELossReduction::None)
```
and for CrossEntropyLoss with reduction `None` they will need to write
```
CrossEntropyLoss(torch::CrossEntropyLossReduction::None)
```
both are significantly longer than their variant version
```
MSELoss(torch::kNone)
```
and
```
CrossEntropyLoss(torch::kNone)
```
Ultimately, we want to make the enum passing experience as similar to Python API as possible, and since we can't use `"none"`, I think writing `torch::kNone` is our second best choice and much better than writing `torch::MSELossReduction::None`.

**Comment by smessmer** [comment]
If you want that, you can just use `enum` instead of `enum class` and you'll get an API where the enum keys are not wrapped in a namespace. But there's a good reason this is discouraged in C++, you lose type safety and the enclosing namespace gets cluttered with keywords. This is why `enum class` is the best practice in C++ these days. However, the variant approach has all the same disadvantages as `enum` and a few more.

**Comment by yf225** [comment]
> you lose type safety

But the "enums" in the variant approach are strongly typed, right? For example, `torch::kNone` is of type `torch::enumtype::kNone`, and it can't be converted to any other type. And if a variant only accepts `torch::enumtype::kNone` type, nothing but an object of `torch::enumtype::kNone` type will be accepted.

> the enclosing namespace gets cluttered with keywords

I don't think this is a concern because we won't be using `torch::kNone` for anything else other than as a "None" keyword constant anyway, so they are already "reserved" in that sense. This PR is just to take advantage of that fact and start to make use of them.

**Comment by ShahriarSS** [comment]
@yf225 The PR looks good. Thanks.

File: torch/csrc/api/include/torch/nn/init.h (hunk 4 above)
Comment (yf225) [reply torch/csrc/api/include/torch/nn/init.h]: @ShahriarSS Thanks a lot for reviewing! I think because the variant type `Nonlinearity` is used in multiple places, it would be better if we declare it in a common place first (e.g. at the top of the file).

File: torch/csrc/api/include/torch/enum.h [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+    otherwise Clang 3.8 would complain:
+    ```
+    error: default initialization of an object of const type 'const enumtype::Enum1'
+    without a user-provided default constructor
+    ```
+  */ \
+  struct k##name { k##name() {} }; \
+} \
+TORCH_API extern const enumtype::k##name k##name; \
+}
```
Comment (yf225) [inline torch/csrc/api/include/torch/enum.h]: Note to self: add test for torch.h include that should include all the enums. add a print test that prints the enum name given token (add operator<< for all enums)

File: c10/util/variant.h [hunk 6]
```diff
@@ -336,7 +336,7 @@ namespace c10 {
   struct in_place_type_t { explicit in_place_type_t() = default; };
 
 #ifdef MPARK_VARIABLE_TEMPLATES
-  constexpr variant_in_place_t in_place{};
+  constexpr variant_in_place_t variant_in_place{};
```
Comment (ezyang) [inline c10/util/variant.h]: What's going on here?

File: torch/csrc/api/include/torch/enum.h [hunk 7]
```diff
@@ -0,0 +1,65 @@
+#pragma once
+
+#include <string>
+
+#include <torch/csrc/WindowsTorchApiMacro.h>
+
+#define TORCH_ENUM_DECLARE(name) \
+namespace torch { \
+namespace enumtype { \
```
Comment (ezyang) [inline torch/csrc/api/include/torch/enum.h]: How'd you decide on this name? Is it just an internal name or is it visible to users?

File: torch/csrc/api/include/torch/enum.h [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+TORCH_API extern const enumtype::k##name k##name; \
+}
+
+#define TORCH_ENUM_DEFINE(name) \
+namespace torch { \
+const enumtype::k##name k##name; \
+}
+
+#define TORCH_ENUM_PRETTY_PRINT(name) \
+std::string operator()(enumtype::k##name& v) const { \
```
Comment (ezyang) [inline torch/csrc/api/include/torch/enum.h]: You can return a `const char*` here right? Seems better.

**Comment by ezyang** [comment]
The way of referencing the enums got a lot wordier in this PR. Was this intentional?

File: c10/util/variant.h (hunk 6 above)
Comment (yf225) [reply c10/util/variant.h]: It's for avoiding conflict with the same variable name in `c10/util/Optional.h`. I added a comment in the c10-specific change log for better clarity.

File: torch/csrc/api/include/torch/enum.h (hunk 7 above)
Comment (yf225) [reply torch/csrc/api/include/torch/enum.h]: It is only an internal name - the type of the enum is e.g. `torch::enumtype::kFanIn`, and the variable that has this type is `torch::kFanIn`, which is what's used by the end user.

File: torch/csrc/api/include/torch/enum.h (hunk 8 above)
Comment (yf225) [reply torch/csrc/api/include/torch/enum.h]: Fixed

**Comment by yf225** [comment]
> The way of referencing the enums got a lot wordier in this PR. Was this intentional?

Yes it was intentional - they will be changed to the short form in https://github.com/pytorch/pytorch/pull/27933.

File: c10/util/variant.h (hunk 6 above)
Comment (ezyang) [reply c10/util/variant.h]: We should just make `variant.h` use the `optional.h` definition. `in_place` is part of the standard.

**Review by ezyang — APPROVED** [review]

File: c10/util/variant.h (hunk 6 above)
Comment (yf225) [reply c10/util/variant.h]: I moved the common definition for `in_place` into `c10/util/in_place.h`, so that both `variant.h` and `optional.h` can use it.
